diff --git a/lib/config/__init__.py b/lib/config/__init__.py
index 6d0a80e..5f5f6dd 100644
--- a/lib/config/__init__.py
+++ b/lib/config/__init__.py
@@ -1011,6 +1011,8 @@ class ConfigWriter(object):
           link = "route:%s" % nic_link
         elif nic_mode == constants.NIC_MODE_OVS:
           link = "ovs:%s" % nic_link
+        elif nic_mode == constants.NIC_MODE_EXT:
+          link = "ext:%s" % nic_link
         else:
           raise errors.ProgrammerError("NIC mode '%s' not handled" % nic_mode)
 
diff --git a/src/Ganeti/Constants.hs b/src/Ganeti/Constants.hs
index a1f84d9..52797c8 100644
--- a/src/Ganeti/Constants.hs
+++ b/src/Ganeti/Constants.hs
@@ -2517,6 +2517,9 @@ nicModeRouted = Types.nICModeToRaw NMRouted
 nicModeOvs :: String
 nicModeOvs = Types.nICModeToRaw NMOvs
 
+nicModeExt :: String
+nicModeExt = Types.nICModeToRaw NMExt
+
 nicIpPool :: String
 nicIpPool = Types.nICModeToRaw NMPool
 
diff --git a/src/Ganeti/Types.hs b/src/Ganeti/Types.hs
index 318127e..6d53d86 100644
--- a/src/Ganeti/Types.hs
+++ b/src/Ganeti/Types.hs
@@ -670,6 +670,7 @@ $(THH.declareLADT ''String "NICMode"
   [ ("NMBridged", "bridged")
   , ("NMRouted",  "routed")
   , ("NMOvs",     "openvswitch")
+  , ("NMExt",     "ext")
   , ("NMPool",    "pool")
   ])
 $(THH.makeJSONInstance ''NICMode)

diff --git a/src/Ganeti/HTools/Backend/IAlloc.hs b/src/Ganeti/HTools/Backend/IAlloc.hs
index 3a67c2d..5379a33 100644
--- a/src/Ganeti/HTools/Backend/IAlloc.hs
+++ b/src/Ganeti/HTools/Backend/IAlloc.hs
@@ -84,6 +84,7 @@
                Just "bridged" -> Ok $ Just Nic.Bridged
                Just "routed" -> Ok $ Just Nic.Routed
                Just "openvswitch" -> Ok $ Just Nic.OpenVSwitch
+               Just "ext" -> Ok Nothing
                Nothing -> Ok Nothing
                _ -> Bad $ "invalid NIC mode in instance " ++ n
   link    <- maybeFromObj a "link"

diff --git a/tools/kvm-ifup.in b/tools/kvm-ifup.in
index 0f7ac44..6a9ff64 100644
--- a/tools/kvm-ifup.in
+++ b/tools/kvm-ifup.in
@@ -46,4 +46,5 @@ if ! is_instance_communication_tap; then
   setup_bridge
   setup_ovs
   setup_route
+  setup_ext
 fi

diff --git a/tools/net-common.in b/tools/net-common.in
index 1ab38e0..390a6c8 100644
--- a/tools/net-common.in
+++ b/tools/net-common.in
@@ -125,3 +125,15 @@ function setup_route {
     fi
   fi
 }
+
+function setup_ext {
+  if [ "$MODE" = "ext" ]; then
+    for NETWORK_TAG in ${NETWORK_TAGS}; do
+      KEY=${NETWORK_TAG//:*/}
+      VAL=${NETWORK_TAG//*:/}
+      if [ "${KEY}" == "ext_provider" -a -x "/usr/share/ganeti/extnetwork/${VAL}/up" ]; then
+         exec "/usr/share/ganeti/extnetwork/${VAL}/up"
+      fi
+    done
+  fi
+}
